### ISSUE TITLE
Cargo doesn't clean build artifacts from old nighlies

### ISSUE THREAD

**Issue by bjorn3** [rust-lang/cargo#5516]
This caused my disk to fill up completely.

**Comment by bjorn3** [comment]
Ping?

**Comment by matklad** [comment]
Cargo itself can't know if a particular artifact will be used in the future or not, because one project could be build using several different Rust versions. So, some explicit user action is required to clean unnecessary artifacts, because only the user knows which artifacts are unnecessary. To that end, `cargo clean` sub command exists. Does it help for your use case?

**Comment by bjorn3** [comment]
Yes, but it is annoying to clean also the compiled dependencies, etc compiled with the current rustc version.

**Comment by Manishearth** [comment]
Can we at least have a global/local cargo config for this, perhaps on by default? I _don't_ compile projects across compilers. I also suspect that's a pretty rare thing for most people to do aside from on CI, and we can ask Travis to flip a config for this.

I just deleted a 140GB target directory, this is kinda ridiculous :smile:

**Comment by Manishearth** [comment]
Another possible heuristic is that we can delete _old_ files that are not currently necessary. So if there are some artifacts from another compiler that were built today, don't touch them, but delete any older ones.

**Comment by bjorn3** [comment]
> So if there are some artifacts from another compiler that were built today, don't touch them, but delete any older ones.

+1 I would like to keep them a bit longer though (~1 or 2 weeks)
